precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+        let signature_bytes: Vec<u8> = signature.into();
+        let account_id = AccountId32::new(account_id.into()).into();
+
+        // 1. Decode the call
+        let call =
+            match Runtime::RuntimeCall::decode_with_depth_limit(DecodeLimit::get(), &mut &*input) {
+                Ok(c) => c,
+                Err(_) => {
+                    log::trace!(target: LOG_TARGET, "Error: could not decode call");
+                    return Ok(false);
+                }
+            };
+
+        // 2. Check that gas limit is not exceeded
+        let info = call.get_dispatch_info();
+        if let Some(gas) = target_gas {
+            let valid_weight = info.weight.ref_time()
+                <= Runtime::GasWeightMapping::gas_to_weight(gas, false).ref_time();
+            if !valid_weight {
+                log::trace!(target: LOG_TARGET, "Error: gas limit exceeded");
+                return Ok(false);
+            }
+        }
+
+        // 3. Recover the ECDSA Public key from the signature
+        let signature_opt = match Self::parse_signature(&signature_bytes) {
+            Some(s) => s,
+            None => {
+                log::trace!(target: LOG_TARGET, "Error: could not parse signature");
+                return Ok(false);
+            }
+        };
+        let payload_hash = Self::build_signing_payload(&account_id);
+        let pubkey = match <pallet_unified_accounts::Pallet<Runtime>>::recover_pubkey(
+            signature_opt.as_ref(),
+            payload_hash,
+        ) {
+            Some(k) => k,
+            None => {
+                log::trace!(
+                    target: LOG_TARGET,
+                    "Error: could not recover pubkey from signature"
+                );
+                return Ok(false);
+            }
+        };
+
+        // 4. Ensure that the caller matches the recovered EVM address from the signature
+        if caller != Self::get_evm_address_from_pubkey(&pubkey) {
+            log::trace!(
+                target: LOG_TARGET,
+                "Error: caller does not match calculated EVM address"
+            );
+            return Ok(false);
+        }
+
+        // 5. Derive the AccountId from the ECDSA compressed Public key
+        let origin = match Self::get_account_id_from_pubkey(pubkey) {
+            Some(a) => a,
+            None => {
+                log::trace!(
+                    target: LOG_TARGET,
+                    "Error: could not derive AccountId from pubkey"
+                );
+                return Ok(false);
+            }
+        };
+
+        // 6. validate the call
+        if let Some(_) = DispatchValidator::validate_before_dispatch(&origin, &call) {
+            return Ok(false);
+        }
+
+        handle
+            .record_external_cost(Some(info.weight.ref_time()), Some(info.weight.proof_size()))?;
```

Comment:
You should do this at the very start, to ensure no free execution in case faulty e.g. signature is submitted.